File: api/src/Services/NotificationService.cs
using Microsoft.EntityFrameworkCore;
using WallotApi.Data;
using WallotApi.Models.DTOs.Notifications;
using WallotApi.Services.Interfaces;

namespace WallotApi.Services;

public class NotificationService : INotificationService
{
    private readonly WallotDbContext _db;

    public NotificationService(WallotDbContext db) => _db = db;

    public async Task<List<NotificationDto>> GetAllAsync(string userId)
    {
        return await _db.Notifications
            .Where(n => n.UserId == userId)
            .OrderByDescending(n => n.CreatedAt)
            .Select(n => new NotificationDto
            {
                Id = n.Id,
                UserId = n.UserId,
                Type = n.Type,
                Title = n.Title,
                Body = n.Body,
                IsRead = n.IsRead,
                CreatedAt = n.CreatedAt
            })
            .ToListAsync();
    }

    public async Task<bool> MarkReadAsync(string userId, string notificationId)
    {
        var notification = await _db.Notifications
            .FirstOrDefaultAsync(n => n.Id == notificationId && n.UserId == userId);
        if (notification == null) return false;

        notification.IsRead = true;
        await _db.SaveChangesAsync();
        return true;
    }

    public async Task MarkAllReadAsync(string userId)
    {
        var unread = await _db.Notifications
            .Where(n => n.UserId == userId && !n.IsRead)
            .ToListAsync();

        foreach (var n in unread)
            n.IsRead = true;

        await _db.SaveChangesAsync();
    }
}

File: services/notificationService.ts
import { api } from '@/lib/api';

export interface Notification {
  id: string;
  user_id: string;
  type: string;
  title: string;
  body: string;
  is_read: boolean;
  created_at: string;
}

export function getNotifications(): Promise<Notification[]> {
  return api.get('/api/notifications');
}

export function markRead(id: string): Promise<void> {
  return api.put(`/api/notifications/${id}/read`, {});
}

export function markAllRead(): Promise<void> {
  return api.put('/api/notifications/read-all', {});
}
